nilearn/connectivity/connectivity_matrices.py

```diff
@@ -0,0 +1,341 @@
+
+    Parameters
+    ----------
+    matrices : list of numpy.ndarray, shape of each (n_features, n_features)
+        List of matrices whose geometric mean to compute. Raise an error if the
+        arrays are not all symmetric positive definite of the same shape.
+
+    init : numpy.ndarray, shape (n_features, n_features) or None, optional
+        Initialization matrix. Raise an error if the array is not symmetric
+        positive definite of the same shape as the elements of matrices. Set to
+        the arithmetic mean of matrices if None.
+
+    max_iter : int, optional (default to 10)
+        Maximal number of iterations.
+
+    tol : float, optional (default to 1e-7)
+        Tolerance.
+
+    Returns
+    -------
+    gmean : numpy.ndarray, shape (n_features, n_features)
+        Geometric mean of the matrices.
+    """
+    # Shape and symmetry positive definiteness checks
+    n_features = matrices[0].shape[0]
+    for matrix in matrices:
+        _check_matrix(matrix, 'square')
+        if matrix.shape[0] != n_features:
+            raise ValueError("Matrices are not of the same shape.")
+        _check_matrix(matrix, 'spd')
+
+    # Initialization
+    matrices = np.array(matrices)
+    if init is None:
+        gmean = np.mean(matrices, axis=0)
+    else:
+        _check_matrix(init, 'square')
+        if init.shape[0] != n_features:
+            raise ValueError("Initialization has incorrect shape.")
+        _check_matrix(init, 'spd')
+        gmean = init
+
+    norm_old = np.inf
+    step = 1.
+
+    # Gradient descent
+    for n in range(max_iter):
+        # Computation of the gradient
+        vals_gmean, vecs_gmean = linalg.eigh(gmean)
+        gmean_inv_sqrt = _form_symmetric(np.sqrt, 1. / vals_gmean, vecs_gmean)
+        whitened_matrices = [gmean_inv_sqrt.dot(matrix).dot(gmean_inv_sqrt)
+                             for matrix in matrices]
+        logs = [_map_eigenvalues(np.log, w_mat) for w_mat in whitened_matrices]
+        logs_mean = np.mean(logs, axis=0)  # Covariant derivative is
+                                           # - gmean.dot(logms_mean)
+        if np.any(np.isnan(logs_mean)):
+            raise FloatingPointError("Nan value after logarithm operation.")
+
+        norm = np.linalg.norm(logs_mean)  # Norm of the covariant derivative on
+                                          # the tangent space at point gmean
+
+        # Update of the minimizer
+        vals_log, vecs_log = linalg.eigh(logs_mean)
+        gmean_sqrt = _form_symmetric(np.sqrt, vals_gmean, vecs_gmean)
+        # Move along the geodesic
+        gmean = gmean_sqrt.dot(
+            _form_symmetric(np.exp, vals_log * step, vecs_log)).dot(gmean_sqrt)
+
+        # Update the norm and the step size
+        if norm < norm_old:
+            norm_old = norm
+        if norm > norm_old:
+            step = step / 2.
+            norm = norm_old
+        if tol is not None and norm / gmean.size < tol:
```

Comment:
again I saw it in group_sparse_covariance.py. I added a similar explanation in the docstring.